src/main/java/org/apache/sysds/runtime/instructions/fed/UnaryMatrixFEDInstruction.java

```diff
@@ -63,17 +72,316 @@ public static UnaryMatrixFEDInstruction parseInstruction(String str) {
+		} else
+			tmp = mo1.getFedMapping().execute(getTID(), true, fr1);
+
+		MatrixObject out = setOutputFedMapping(ec, mo1, fr1.getID());
+
+		if(mo1.isFederated(FederationMap.FType.ROW)) {
+			aggResults(ec, mo1, tmp, out);
+		}
+	}
+
+	private void aggResults(ExecutionContext ec, MatrixObject mo1, Future<FederatedResponse>[] tmp, MatrixObject out) {
+		NewVariable tmpVar;
+		if(Arrays.asList("ucumk+", "ucum*").contains(getOpcode())) {
+			tmpVar =  getSumOrProdVariable(ec, mo1, tmp, getOpcode().equals("ucumk+"));
+			aggPartialResults(out, tmpVar);
+			ec.removeVariable(String.valueOf(tmpVar._id));
+		}
+		else if(Arrays.asList("ucummin", "ucummax").contains(getOpcode())) {
+			tmpVar = getMinOrMaxVariable(ec, mo1, tmp, getOpcode().equals("ucummin"));
+			aggPartialResults(out, tmpVar);
+			ec.removeVariable(String.valueOf(tmpVar._id));
+		} else {
+			MatrixBlock scalingValues = computeScalingValues(ec, mo1, tmp);
+			NewVariable[] ternaryVars = getTernaryVars(ec, mo1, scalingValues);
+			setScalingValues(ec, mo1, out,ternaryVars);
+			aggSumprod(out);
+		}
+	}
+
+	private void aggPartialResults(MatrixObject out, NewVariable tmpVar) {
+		CPOperand operand = new CPOperand(String.valueOf(tmpVar._id), ValueType.FP64, DataType.MATRIX);
+		modifyInstString(operand);
+
+		FederatedRequest[] fr1 = out.getFedMapping().broadcastSliced(tmpVar._mo, false);
+		FederatedRequest fr2 = FederationUtils.callInstruction(instString, output, out.getFedMapping().getID(),
+			new CPOperand[]{output, operand},
+			new long[]{out.getFedMapping().getID(), fr1[0].getID()});
+		FederatedRequest fr3 = out.getFedMapping().cleanup(getTID(), fr1[0].getID());
+		out.getFedMapping().execute(getTID(), true, fr1, fr2, fr3);
+
+		out.setFedMapping(out.getFedMapping().copyWithNewID(fr2.getID()));
+	}
+
+	private NewVariable getMinOrMaxVariable(ExecutionContext ec, MatrixObject mo1, Future<FederatedResponse>[] tmp, boolean isMin) {
+		MatrixBlock val = new MatrixBlock(1, (int) mo1.getNumColumns(), isMin ? Double.MAX_VALUE : -Double.MAX_VALUE);
+
+		int size = (int)mo1.getNumRows();
+		MatrixBlock retBlock = new MatrixBlock(size, (int)mo1.getNumColumns(), isMin ? Double.MAX_VALUE : -Double.MAX_VALUE);
+		BinaryOperator bop = InstructionUtils.parseBinaryOperator(isMin ? "min" : "max");
+
+		for(int i = 0; i < tmp.length - 1; i++)
+			try {
+				MatrixBlock curr = ((MatrixBlock) tmp[i].get().getData()[0]);
+				curr = curr.slice(curr.getNumRows()-1,curr.getNumRows()-1);
+
+				val = curr.binaryOperationsInPlace(bop, val);
+				size = (int) (mo1.getNumRows() - mo1.getFedMapping().getFederatedRanges()[i].getEndDims()[0]);
+				MatrixBlock mb = new MatrixBlock(size, (int) mo1.getNumColumns(), 0.0)
+					.binaryOperations(InstructionUtils.parseBinaryOperator("+"), val, new MatrixBlock());
+
+				int from = (int) mo1.getFedMapping().getFederatedRanges()[i+1].getBeginDims()[0];
+				int to = (int) mo1.getFedMapping().getFederatedRanges()[tmp.length-1].getEndDims()[0]-1;
```

Comment:
why does it always cover the range of all rows from current position to the end of the federated matrix?